tests/integration/azure/test_azure_credentials.py

```diff
@@ -0,0 +1,142 @@
+import base64
+import re
+from unittest.mock import MagicMock
+
+import pytest
+from databricks.labs.blueprint.tui import MockPrompts
+
+from databricks.labs.ucx.assessment.azure import AzureServicePrincipalInfo
+from databricks.labs.ucx.assessment.crawlers import _SECRET_PATTERN
+from databricks.labs.ucx.azure.access import StoragePermissionMapping
+from databricks.labs.ucx.azure.azure_credentials import AzureServicePrincipalMigration
+
+
+@pytest.fixture
+def prepare_spn_migration_test(ws, debug_env, make_random):
+    def inner(read_only=False) -> dict:
```

Comment:
1) don't return `dict` - it's broken
2) why do you need `inner`? do you cleanup anything you've created here?

rewrite.

Reply:
1. replaced with a dataclass. Also separated the logic of fetching the test info like application_id etc. and creating parameters for ServicePrincipalMigration object to run the migration test.
2. `inner` is easier for passing in parameters with different round of tests. Here we do not create anything in the lab-azure env.